### PR TITLE
Filterbar Revamp

### PR DESCRIPTION
**Pull request by jdtoombs**
New filter bar revamp, clicking Find available properties will allow you to search by:

- location
- project number
- Building predominate use
- Building construction type
- Building number of floors
- Net rentable area
- Min/max lot size
- Bare land only

See GIF below for UI:
_**Updated**: Bare Land Only filter now filters out land with buildings on it properly and changed building name -> name so users can search for any property name_
fitler

### CODE REVIEW

File: backend/api/Areas/Property/Models/Search/GeoJsonFilterModel.cs [hunk 1]
```diff
@@ -74,6 +74,12 @@ public class GeoJsonFilterModel : PageFilter
         /// <value></value>
         public bool? IgnorePropertiesInProjects { get; set; }
 
+        /// <summary>
```
Comment (jdtoombs) [inline backend/api/Areas/Property/Models/Search/GeoJsonFilterModel.cs:77]: Realized that this may be redundant, should I just send `PropertyType.Land` as the `PropertyType` filter instead of creating a new flag?

File: backend/api/Areas/Property/Models/Search/PropertyFilterModel.cs [hunk 2]
```diff
@@ -73,6 +73,12 @@ public class PropertyFilterModel : PageFilter
         /// <value></value>
         public bool? IgnorePropertiesInProjects { get; set; }
 
+        /// <summary>
+        /// get/set - Flag indicating bare land is the desired return.
+        /// </summary>
+        /// <value></value>
+        public bool? BareLandOnly { get; set; }
```
Comment (jdtoombs) [inline backend/api/Areas/Property/Models/Search/PropertyFilterModel.cs:80]: See above comment

File: frontend/src/components/SearchBar/FindMorePropertiesForm.tsx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    display: flex;
+    .form-label {
+      margin-top: 5px;
+      margin-right: 10px;
+      width: 150px;
+      text-align: right;
+    }
+  }
+`;
+
```
Comment (jdtoombs) [inline frontend/src/components/SearchBar/FindMorePropertiesForm.tsx:25]: Will add some documentation to these styled components

File: frontend/src/components/common/form/ParentSelect.tsx [hunk 4]
```diff
@@ -55,6 +55,21 @@ export const ParentSelect: React.FC<IParentSelect> = ({
   /** controls the multi selections displayed to the user */
   const [multiSelections, setMultiSelections] = React.useState<any>([]);
 
+  const headerAbbrvCheck = (options: SelectOptions, cutOffLength: number, parent: any) => {
```
Comment (jdtoombs) [inline frontend/src/components/common/form/ParentSelect.tsx:58]: This is going to be removed and we are going to adjust the size of the options menu to contain all the agency labels, just going to to do that as a separate PR

File: frontend/src/components/maps/FindMorePropertiesButton.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      </Popover.Content>
+    </StyledPopover>
+  );
+
+  return (
+    <OverlayTrigger trigger="click" rootClose transition overlay={popover} placement="bottom">
+      <Button style={{ height: 38 }}>
+        <ButtonContent>
+          <FaSign size={20} style={{ marginRight: 5, marginTop: 3 }} />
+          {window.screen.width > 1080 && <p>Find available properties</p>}
```
Comment (jdtoombs) [inline frontend/src/components/maps/FindMorePropertiesButton.tsx:57]: Had this as I had the app up on my portrait monitor for a while... but it does fit it kind of nicely this way.

File: frontend/src/components/maps/FindMorePropertiesButton.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        <TitleContent />{' '}
+      </Popover.Title>
+      <Popover.Content>
+        <FindMorePropertiesForm />
+      </Popover.Content>
+    </StyledPopover>
+  );
+
+  return (
+    <OverlayTrigger trigger="click" rootClose transition overlay={popover} placement="bottom">
```
Comment (jdtoombs) [inline frontend/src/components/maps/FindMorePropertiesButton.tsx:53]: Right now it closes via the 'x' button and clicking outside of the form. But I have a feeling this isn't desired behaviour. Currently adjusting, going to confirm with Bobbi.

File: frontend/src/features/properties/filter/PropertyFilter.test.tsx [hunk 7]
```diff
@@ -182,12 +183,10 @@ describe('MapFilterBar', () => {
...
       minLotSize: '10',
       maxLotSize: '20',
-      inSurplusPropertyProgram: 'true',
+      inSurplusPropertyProgram: TrueFalse.True,
     };
-    const { getByText, getByPlaceholderText } = render(getUiElement(providedFilter));
+    const { getByText } = render(getUiElement(providedFilter));
     expect(getByText('Address')).toBeVisible();
     expect(getByText('Core Operational')).toBeVisible();
-    expect(getByPlaceholderText('Min Lot Size')).toHaveValue('10');
```
Comment (jdtoombs) [inline frontend/src/features/properties/filter/PropertyFilter.test.tsx:190]: Will adjust these tests after getting functional tasks done

File: frontend/src/features/properties/filter/PropertyFilterOptions.tsx [hunk 8]
```diff
@@ -9,16 +9,12 @@ import { IPropertyFilter } from './IPropertyFilter';
 export const PropertyFilterOptions: React.FC = () => {
   const state: { options: any[]; placeholders: Record<string, string> } = {
     options: [
+      { label: 'Building Name', value: 'buildingName' },
       { label: 'Address', value: 'address' },
-      { label: 'Location', value: 'administrativeArea' },
```
Comment (jdtoombs) [inline frontend/src/features/properties/filter/PropertyFilterOptions.tsx:13]: Wanted to confirm this was correct to remove since we have it in the new filter?

File: backend/api/Areas/Property/Models/Search/GeoJsonFilterModel.cs (hunk 1 above)
Comment (Fosol) [reply backend/api/Areas/Property/Models/Search/GeoJsonFilterModel.cs:77]: I think it already has the `PropertyType` filter.  You should just use it instead of a duplicate filter that does less.

File: backend/api/Areas/Property/Models/Search/GeoJsonFilterModel.cs [hunk 9]
```diff
@@ -156,6 +162,12 @@ public class GeoJsonFilterModel : PageFilter
         /// <value></value>
         public int? ConstructionTypeId { get; set; }
 
+        /// <summary>
+        /// get/set - Building  name.
+        /// </summary>
+        /// <value></value>
+        public string? BuildingName { get; set; }
```
Comment (Fosol) [inline backend/api/Areas/Property/Models/Search/GeoJsonFilterModel.cs:169]: Just filter on `Name`, then it'll work for either land or building.

Comment (Fosol) [reply backend/api/Areas/Property/Models/Search/GeoJsonFilterModel.cs:169]: Also remove the `nullable`, as we're not using this new feature presently.  `string` are already nullable by nature.

File: backend/api/Areas/Property/Models/Search/PropertyFilterModel.cs (hunk 2 above)
Comment (Fosol) [reply backend/api/Areas/Property/Models/Search/PropertyFilterModel.cs:80]: Remove and use the `PropertyType` instead.

File: backend/api/Areas/Property/Models/Search/PropertyFilterModel.cs [hunk 10]
```diff
@@ -86,6 +92,12 @@ public class PropertyFilterModel : PageFilter
         /// <value></value>
         public bool? InEnhancedReferralProcess { get; set; }
 
+        /// <summary>
+        /// get/set - Value of the building name to be filtered.
+        /// </summary>
+        /// <value></value>
+        public string BuildingName { get; set; }
```
Comment (Fosol) [inline backend/api/Areas/Property/Models/Search/PropertyFilterModel.cs:99]: Rename to `Name` so that it can be used by land and building.

File: backend/entities/Models/BuildingFilter.cs [hunk 11]
```diff
@@ -81,11 +81,11 @@ public BuildingFilter(double neLat, double neLong, double swLat, double swLong)
         /// <param name="maxAssessedValue"></param>
         /// <param name="sort"></param>
         /// <returns></returns>
-        public BuildingFilter(string address, int? agencyId, int? constructionTypeId, int? predominantUseId, int? floorCount, string tenancy, float? minRentableArea, float? maxRentableArea, decimal? minMarketValue, decimal? maxMarketValue, decimal? minAssessedValue, decimal? maxAssessedValue, string[] sort)
+        public BuildingFilter(string address, int? agencyId, int? constructionTypeId, int? predominateUseId, int? floorCount, string tenancy, float? minRentableArea, float? maxRentableArea, decimal? minMarketValue, decimal? maxMarketValue, decimal? minAssessedValue, decimal? maxAssessedValue, string[] sort)
         {
             this.Address = address;
             this.ConstructionTypeId = constructionTypeId;
-            this.PredominateUseId = predominantUseId;
+            this.PredominateUseId = predominateUseId;
```
Comment (Fosol) [inline backend/entities/Models/BuildingFilter.cs:88]: Thanks for fixing

File: backend/entities/Models/PropertyFilter.cs [hunk 12]
```diff
@@ -47,6 +47,11 @@ public abstract class PropertyFilter : PageFilter
         /// </summary>
         public string ProjectNumber { get; set; }
 
+        /// <summary>
+        /// get/set - Flag for only returning bare land.
+        /// </summary>
+        public bool? BareLandOnly { get; set; }
```
Comment (Fosol) [inline backend/entities/Models/PropertyFilter.cs:53]: Use `PropertyType` filter instead.

File: backend/entities/Models/PropertyFilter.cs [hunk 13]
```diff
@@ -65,11 +70,17 @@ public abstract class PropertyFilter : PageFilter
         /// <value></value>
         public bool? InEnhancedReferralProcess { get; set; }
 
+        /// <summary>
+        /// get/set - The value of the building name.
+        /// </summary>
+        /// <value></value>
+        public string BuildingName { get; set; }
```
Comment (Fosol) [inline backend/entities/Models/PropertyFilter.cs:77]: Replace with `Name` so it will be used for land and building.

File: frontend/src/components/common/form/Input.tsx [hunk 14]
```diff
@@ -36,6 +37,8 @@ type OptionalAttributes = {
   tooltip?: string;
   /** Display errors in a tooltip instead of in a div */
   displayErrorTooltips?: boolean;
+  /** add inline style to the input component */
+  style?: CSSProperties;
```
Comment (Fosol) [inline frontend/src/components/common/form/Input.tsx:41]: Neat

File: frontend/src/components/maps/leaflet/InventoryLayer.tsx [hunk 15]
```diff
@@ -80,6 +80,12 @@ export const InventoryLayer: React.FC<InventoryLayerProps> = ({
       maxLandArea: filter?.maxLandArea,
       inSurplusPropertyProgram: filter?.inSurplusPropertyProgram,
```
Comment (Fosol) [inline frontend/src/components/maps/leaflet/InventoryLayer.tsx:81]: With so many properties it feels like it should just accept the filter object itself instead.

File: frontend/src/components/maps/leaflet/Map.tsx [hunk 16]
```diff
@@ -77,7 +77,7 @@ export type LayerPopupInformation = PopupContentConfig & {
 };
 
 const defaultFilterValues: IPropertyFilter = {
-  searchBy: 'address',
+  searchBy: 'buildingName',
```
Comment (Fosol) [inline frontend/src/components/maps/leaflet/Map.tsx:80]: Change to `name`

File: frontend/src/features/properties/filter/PropertyFilter.test.tsx [hunk 17]
```diff
@@ -182,12 +183,10 @@ describe('MapFilterBar', () => {
       classificationId: '0',
       minLotSize: '10',
       maxLotSize: '20',
-      inSurplusPropertyProgram: 'true',
+      inSurplusPropertyProgram: TrueFalse.True,
```
Comment (Fosol) [inline frontend/src/features/properties/filter/PropertyFilter.test.tsx:186]: Ha see, they're so much better than magic strings.

File: frontend/src/components/SearchBar/FindMorePropertiesForm.tsx [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+          <NumberInput label="Lot size" field="minLotSize" placeholder="min" />
+          <span style={{ marginTop: 5, marginLeft: 5, marginRight: 5 }}>-</span>
+          <NumberInput field="maxLotSize" placeholder="max" />
+          <VerticalLine />
+          <Check label="Bare land only" field="bareLandOnly" />
+        </StyledRow>
+      </FormSection>
+      <Row>
+        <SearchButton onClick={handleSubmit}>Search</SearchButton>
+      </Row>
```
Comment (agalbrai) [inline frontend/src/components/SearchBar/FindMorePropertiesForm.tsx:127]: is there a way you could have the search form close on "Search" so that it wasn't blocking the map anymore?

Comment (jdtoombs) [reply frontend/src/components/SearchBar/FindMorePropertiesForm.tsx:127]: Yup this is possible, just wondering if they would get confused if it is closed right away though.. think I will leave as is for now and there can be a new UX ticket after feedback

File: frontend/src/features/properties/filter/PropertyFilter.test.tsx (hunk 17 above)
Comment (jdtoombs) [reply frontend/src/features/properties/filter/PropertyFilter.test.tsx:186]: Haha the magic strings will slowly disappear...

File: frontend/src/components/maps/FindMorePropertiesButton.tsx (hunk 5 above)
Comment (devinleighsmith) [reply frontend/src/components/maps/FindMorePropertiesButton.tsx:57]: probably should do this with @media queries instead.

Comment (jdtoombs) [reply frontend/src/components/maps/FindMorePropertiesButton.tsx:57]: Ooh thank you, I'll read up on them and adjust

**Comment by devinleighsmith** [comment]
looks like this review is still waiting for updates, will review once the original comments are addressed.

File: backend/dal/Helpers/Extensions/PropertyExtensions.cs [hunk 19]
```diff
@@ -43,7 +43,15 @@ public static IQueryable<Entity.Views.Property> GenerateQuery(this PimsContext c
                 query = query.Where(p => !p.IsSensitive);
 
             if (filter.PropertyType.HasValue)
-                query = query.Where(p => p.PropertyTypeId == filter.PropertyType);
+                query = query.Where(p => p.PropertyTypeId == filter.PropertyType.Value);
+
+            if (filter.BareLandOnly == true)
```
Comment (jdtoombs) [inline backend/dal/Helpers/Extensions/PropertyExtensions.cs:48]: Seems to work, but want confirmation that this is ok!

File: backend/api/Areas/Property/Controllers/SearchController.cs [hunk 20]
```diff
@@ -121,6 +121,7 @@ public IActionResult GetGeoJson([FromBody] GeoJsonFilterModel filter)
             if (!filter.IsValid()) throw new BadRequestException("Property filter must contain valid values.");
 
             var pfilter = filter.CopyValues(new AllPropertyFilter());
+            pfilter.PropertyType = filter.PropertyType;
```
Comment (jdtoombs) [inline backend/api/Areas/Property/Controllers/SearchController.cs:124]: Tried doing `mapper.Map<AllPropertyFilter>(filter)` but the value still wasn't being sent over. This did allow it to work; however, not sure if it is necessary anymore but thought I would leave it in in case we need to use this filter eventually. Or should I just take it out until we get the mapper/copyvals working?

Comment (Fosol) [reply backend/api/Areas/Property/Controllers/SearchController.cs:124]: Ah, I'll take a look at this if I find time in a week or so.

File: backend/dal/Helpers/Extensions/PropertyExtensions.cs (hunk 19 above)
Comment (Fosol) [reply backend/dal/Helpers/Extensions/PropertyExtensions.cs:48]: Looks correct, the real test is the result :P

File: backend/entities/Models/AllPropertyFilter.cs [hunk 21]
```diff
@@ -110,14 +109,15 @@ public AllPropertyFilter(Dictionary<string, Microsoft.Extensions.Primitives.Stri
...
-            this.PropertyType = Enum.TryParse(typeof(PropertyTypes), filter.GetStringValue(nameof(this.PropertyType), null), out object propType) ? (PropertyTypes?)propType : null;
+            PropertyTypes propType;
 
             this.PID = filter.GetStringValue(nameof(this.PID));
             this.Zoning = filter.GetStringValue(nameof(this.Zoning));
             this.ZoningPotential = filter.GetStringValue(nameof(this.ZoningPotential));
             this.MinLandArea = filter.GetFloatNullValue(nameof(this.MinLandArea));
             this.MaxLandArea = filter.GetFloatNullValue(nameof(this.MaxLandArea));
 
+            this.PropertyType = Enum.TryParse(filter.GetStringValue(nameof(this.PropertyType), null), out propType) ? (PropertyTypes?)propType : null;
```
Comment (Fosol) [inline backend/entities/Models/AllPropertyFilter.cs:120]: I think you can do this in one line like before using the new syntax.

File: frontend/src/components/maps/FindMorePropertiesButton.tsx [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+/** icon for sign used within component, styled blue */
+const TitleForSaleSign = styled(FaSign)`
+  fill: #003366;
+  margin-right: 5px;
+`;
+
+/** this component contains the trigger for additional filter options off the base filter */
+export const FindMorePropertiesButton: React.FC<any> = () => {
+  const smallScreen = useMediaQuery({ maxWidth: 1800 });
+  const TitleContent = () => (
```
Comment (Fosol) [inline frontend/src/components/maps/FindMorePropertiesButton.tsx:36]: Having these functions like this will result in them being created each time the component is run.  Ideally you would `useCallback` or `useMemo`.   Or you would extract as a separate component.

File: frontend/src/components/maps/FindMorePropertiesButton.tsx [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+    <div style={{ display: 'flex' }}>
+      <TitleForSaleSign size={42} />
+      <h3 style={{ color: '#003366' }}>Find available surplus properties</h3>
+      <CloseButton onClick={() => document.body.click()} />
+    </div>
+  );
+
+  /** this provides a way to create a form with tooltip like behaviour in the overlay trigger */
+  const popover = (
```
Comment (Fosol) [inline frontend/src/components/maps/FindMorePropertiesButton.tsx:45]: Having these functions like this will result in them being created each time the component is run.  Ideally you would `useCallback` or `useMemo`.   Or you would extract as a separate component.

**Review by Fosol — APPROVED** [review]

File: frontend/src/components/maps/FindMorePropertiesButton.tsx (hunk 22 above)
Comment (jdtoombs) [reply frontend/src/components/maps/FindMorePropertiesButton.tsx:36]: Oh good point, I can adjust that!

**Review by agalbrai — COMMENTED** [review]
Looking forward to it 👍

File: backend/api/Areas/Property/Models/Search/GeoJsonFilterModel.cs [hunk 24]
```diff
@@ -249,7 +255,8 @@ public bool IncludeBuildings
                     || this.FloorCount.HasValue
                     || !String.IsNullOrWhiteSpace(this.Tenancy)
                     || this.MinRentableArea.HasValue
-                    || this.MaxRentableArea.HasValue;
+                    || this.MaxRentableArea.HasValue
+                    || this.RentableArea.HasValue;
```
Comment (agalbrai) [inline backend/api/Areas/Property/Models/Search/GeoJsonFilterModel.cs:259]: Is this for "net" rentable area?
